File: src/day1/BTPreOrder.go
package main

func PreOrderSearchWalk(current *BinaryNode, path []int) []int {
	if current == nil {
		return path
	}

	// recurse
	// pre

	// recurse
	path = append(path, current.Value)

	path = PreOrderSearchWalk(current.Left, path)

	path = PreOrderSearchWalk(current.Right, path)

	// post
	return path
}

//           7             <- root / head
//         /   \               |
//       23      3             |
//     /   \   /   \           |
//    5    4   18   21     <- leaf

// pre_order_search => [7, 23, 5, 4, 3, 18, 21]
func PreOrderSearch(head *BinaryNode) []int {
	return PreOrderSearchWalk(head, []int{})
}

File: src/day1/BTPreOrder.ts
function walk(current: BinaryNode<number> | null, path: number[]): number[]{
    if (!current) {
        return path
    }

    // recurse
    // pre
    
    // recurse
    path.push(current.value)

    walk(current.left, path)

    walk(current.right, path)
    
    // post
    return path
}

//    Depth First Search / Traversal
//    using implicit Data Structure -> Stack
// 
//           7             <- root / head
//         /   \               |
//       23      3             |
//     /   \   /   \           |
//    5    4   18   21     <- leaf
// 
//    pre_order_search => [7, 23, 5, 4, 3, 18, 21]
export default function pre_order_search(head: BinaryNode<number>): number[] {
    return walk(head, [])
}